app.go

```diff
@@ -47,29 +49,31 @@ type optionFunc func(*App)
 
 func (f optionFunc) apply(app *App) { f(app) }
 
-// Provide registers constructors with the application's dependency injection
-// container. Constructors provide one or more types, can depend on other
-// types available in the container, and may optionally return an error. For
-// example:
+// Provide registers any number of constructor functions, teaching the
+// application how to instantiate various types. The supplied constructor
+// functions must return one or more objects, may depend on other types
+// available in the application, and may return an error. For example:
```

Comment:
consider: might be best to lead in with "may depend on other types..." so that the notes concerning the return types are consecutive. i.e:

"The supplied constructor functions may depend on other types available in the application, must return one or more objects, and may return an error."

I realize you may have thought of this and ordered with "must" followed by "may" notes, but felt it was worth suggesting anyways. I'm fine with either.